githubapp/scheduler.go

```diff
@@ -0,0 +1,238 @@
+// Copyright 2020 Palantir Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package githubapp
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/pkg/errors"
+	"github.com/rcrowley/go-metrics"
+	"github.com/rs/zerolog"
+)
+
+const (
+	MetricsKeyQueueLength   = "github.event.queued"
+	MetricsKeyActiveWorkers = "github.event.workers"
+)
+
+var (
+	ErrCapacityExceeded = errors.New("scheduler: capacity exceeded")
+)
+
+// Dispatch is a webhook payload and the handler that handles it.
+type Dispatch struct {
+	Handler EventHandler
+	Ctx     context.Context
+
+	EventType  string
+	DeliveryID string
+	Payload    []byte
+}
+
+// Execute calls the Dispatch's handler with the stored arguments.
+func (d Dispatch) Execute() error {
+	return d.Handler.Handle(d.Ctx, d.EventType, d.DeliveryID, d.Payload)
+}
+
+// AsyncErrorCallback is called by an asynchronous scheduler when an event
+// handler returns an error. The error from the handler is passed directly as
+// the final argument.
+type AsyncErrorCallback func(ctx context.Context, err error)
+
+// DefaultAsyncErrorCallback logs errors.
+func DefaultAsyncErrorCallback(ctx context.Context, err error) {
+	zerolog.Ctx(ctx).Error().Err(err).Msg("Unexpected error handling webhook")
+}
+
+// ContextDeriver creates a new independent context from a request's context.
+// The new context must be based on context.Background(), not the input.
+type ContextDeriver func(context.Context) context.Context
```

Comment:
why not have the Scheduler decide how to derive the context? then we could have

```
Scheduler.Schedule(ctx context.Context, dispatch Dispatch)
```

it feels like context deriving should be tied explicitly to schedule logic, rather than be swappable

Reply:
Context derivation is customizable because only the end user can know the context values that are relevant to copy to the new context for the handler, particularly if they've added custom middleware or other stuff before the event dispatcher. This felt like the right tradeoff to me, rather than having user implement their own schedulers to change the context, but maybe there are more elegant primitives to expose.

Regarding the signature, I was lazy: `Dispatch` contains the context so I could reuse the type in the internal queue for `QueueAsysnScheduler`. Looking at it now, I should probably make the `context.Context` a function argument and define a new private type for the queue.